---
title: "Cellranger mapping QC"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
## Objective
In this Rmarkdown document, we are going to perform a quality control (QC) analysis of the mapping results obtained by running [cellranger version 5.0.1]("https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger").

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(gt)
library(ggpubr)
library(ggrepel)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r warning = FALSE, message = FALSE}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/CLARIAJOA_metadata.csv")
path_cellranger_metrics_gex <- here::here("01_cellranger_mapping/results/cellranger_mapping_metrics_count.csv")

# Functions
source(here::here("bin/utils.R"))
```

## Load data
The data used in this Rmarkdown document comes from **CLARIAJOA** dataset, generated with a Cell Hashing approach.
```{r}
# Load metadata
metadata <- read.csv(path_project_metadata)
metadata <- metadata[order(metadata$gem_id), !names(metadata) %in% c("project", "subproject"), drop = F]

# Get GEM ID list
hashed_gemid_list <- sort(unique(metadata$gem_id[metadata$hashing == "hashed"]))
not_hashed_gemid_list <- sort(unique(metadata$gem_id[metadata$hashing == "not_hashed"]))

# Load cellranger metrics file [GEX + HTO]
cellranger_metrics_gex <- read_csv(path_cellranger_metrics_gex, col_names = TRUE)
cellranger_metrics_gex <- cellranger_metrics_gex[order(cellranger_metrics_gex$gem_id), ]
metadata_to_merge <- subset(metadata, type == "cDNA")
cellranger_metrics_gex <- merge(metadata_to_merge, cellranger_metrics_gex)

# Clean metrics dataframe
cellranger_metrics_gex_df <- clean_cellranger_metrics_df(cellranger_metrics_gex)

# Split samples metrics by hashing library type
metrics_gex_df <- group_hashing_cellranger_metrics_df(
  cellranger_metrics_gex_df, not_hashed_gemid_list, hashed_gemid_list)
```

# Data overview
```{r}
# Print loaded data
print("Libraries metadata")
metadata

print("GEX QC summary table")
metrics_gex_df
```

# Gene Expression QC
We will start by showing the three most relevant metrics (number of reads, estimated number of recovered cells, fraction of reads in cells, mean reads per cell, fraction of reads mapped to exonic reads, and median genes per cell) obtained by cellranger for each of the working libraries. This information will give us an idea of the quality of the experiment as well as the sequencing and the mapping steps.

```{r fig.width=14}
table_cellranger_metrics_gex_hash(metrics_gex_df = metrics_gex_df,
                             cellranger_version = "CellRanger v6.1.1")
```


## Sequencing QC
First, we will assess the quality of the sequenced libraries before the mapping step. To so do, we will use the "Q30" variables in our dataset, as these report the fraction of the bases with a Q-score >30 for different sequences (barcodes, reads and UMIs).
Q-score is calculated in the following way:
$$Q = -10\log10(p)$$
where *p* is the probability of the base being wrongly called. Thus, bases with a high Q-score are reliable.

```{r fig.height = 6, fig.width = 14}
qc_seq_vars <- c("Q30_Bases_in_Barcode",
                 "Q30_Bases_in_RNA_Read",
                 "Q30_Bases_in_UMI")

gg_qc_seq <- purrr::map(qc_seq_vars, function(var) {
  ggplot(metrics_gex_df, aes_string(x = "library_name", y = var, fill = "gem_id")) +
    geom_col() +
    facet_grid(~hashing, scales = "free_x", space = "free") +
    theme_bw() +
    
    ylim(0, 100) +
    labs(title = str_replace_all(var, "_", " "),
         x = "Libraries (GEM IDs)",
         y = "Q30 Bases (%)") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 8),
          axis.text.x = element_text(hjust = 1, angle = 45),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")
})

ggarrange(plotlist = gg_qc_seq, nrow = 1, ncol = 3)
```

## Mapping QC
Next, we will check the quality of the mapping step performed by `cellranger 5.0.0` across libraries. To do so, we will compare the percentage of reads mapped to the genome, and within these mapped reads, the amount of reads mapped to intergenic regions, intronic and exonic regions. We aim to obtain libraries with a high percentage of confidently mapped reads, and specially a high percentage of exonic reads, which correspond with gene expression or RNAs. The reads mapping to intergenic regions suggest contamination of ambient DNA, whereas reads mapping to intronic regions may come from pre-mRNAs or mature spliced isoforms that retain certain introns.

```{r fig.height = 12, fig.width = 14}
qc_map_vars <- c("Reads_Mapped_Confidently_to_Genome",
                 "Reads_Mapped_Confidently_to_Intergenic_Regions",
                 "Reads_Mapped_Confidently_to_Intronic_Regions",
                 "Reads_Mapped_Confidently_to_Exonic_Regions")

gg_qc_map <- purrr::map(qc_map_vars, function(var) {
  ggplot(metrics_gex_df, aes_string(x = "library_name", y = var, fill = "gem_id")) +
    geom_col() +
    facet_grid(~hashing, scales = "free_x", space = "free") +
    theme_bw() +
    
    ylim(0, 100) +
    labs(title = str_replace_all(var, "_", " "),
         x = "Libraries (GEM IDs)",
         y = "Reads Mapped (%)") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          axis.text.x = element_text(hjust = 1, angle = 45),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")
})

ggarrange(plotlist = gg_qc_map, nrow = 2, ncol = 2)
```

## Sequencing saturation and depth
After assessing mapped reads, it is important to test which is the sequencing saturation and depth for each library. The sequencing saturation is dependent on the library complexity and sequencing depth. The library complexity is the total number of different transcripts present in the library and it varies between the cell types/tissues, whereas the sequencing depth is the number of Set2 reads per cell. For this reason, we will plot the number of detected genes as a function of depth (sequenced reads). As sequencing depth increases, more genes are detected, but this function reaches a plateau, whereby more sequenced reads does not result in more detected genes; therefore, at this point we assure we sequenced until saturation. More specifically, the sequencing saturation the fraction of confidently mapped, valid cell-barcode, valid UMI reads that had a non-unique (cell-barcode, UMI, gene).

```{r fig.height = 12, fig.width = 14}
gg_lib_size <- metrics_gex_df %>%
  mutate(Number_of_Reads_mil = Number_of_Reads / 1000000) %>%
  ggplot(aes(x = library_name, y = Number_of_Reads_mil, fill = gem_id)) +
  geom_bar(stat = "identity") +
  facet_grid(~hashing, scales = "free_x", space = "free") +
  theme_bw() +
  
  labs(title = "Library size",
    x = "Libraries (GEM IDs)", 
    y =  "Number Reads (in millions)") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 8),
          axis.text.x = element_text(hjust = 1, angle = 45),
        strip.placement = "outside",
        strip.background = element_rect(colour = NA),
        legend.position = "none")

gg_qc_seq_sat <- metrics_gex_df %>%
  mutate(Sequencing_Saturation_perc = Sequencing_Saturation / 100,
         Mean_Reads_per_Cell_tho = Mean_Reads_per_Cell / 1000) %>%
  ggplot(aes(x = Mean_Reads_per_Cell_tho,
             y = Sequencing_Saturation_perc, color = gem_id)) +
  geom_point() +
  theme_bw() +
  
  ylim(0, 1) +
  geom_text_repel(aes(label = library_name), size = 4) +
  labs(title = "Sequencing Saturation",
       x = "Mean Reads per Cell (in thousands)", 
       y = "") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.position = "none")

gg_qc_seq_depth_cell <- metrics_gex_df %>%
  mutate(Mean_Reads_per_Cell_tho = Mean_Reads_per_Cell / 1000) %>%
  ggplot(aes(x = Mean_Reads_per_Cell_tho,
             y = Median_Genes_per_Cell, color = gem_id)) +
  geom_point() +
  theme_bw() +
  
  geom_text_repel(aes(label = library_name), size = 4) +
  labs(x = "Mean Reads per Cell (in thousands)",
       y = "Mean Detected Genes per Cell") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.position = "none")

gg_qc_seq_depth <- metrics_gex_df %>%
  mutate(Number_of_Reads_mil = Number_of_Reads / 1000000) %>%
  ggplot(aes(x = Number_of_Reads_mil,
             y = Total_Genes_Detected, color = gem_id)) +
  geom_point() +
  theme_bw() +
  
  geom_text_repel(aes(label = library_name), size = 4) +
  labs(x = "Number of Reads (in millions)", 
       y = "Total Genes Detected") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.position = "none")

cowplot::plot_grid(gg_lib_size, gg_qc_seq_sat, 
                   gg_qc_seq_depth_cell, gg_qc_seq_depth, 
                   nrow = 2, ncol = 2)
```


# Cell Hashing QC (antibody capture)
In this project we followed a Cell Hashing approach using Cell Surface Proteins (CSP) for overloading, only in one sample. Therefore, we aim to analyze both the antibody usage and the sequencing depth of the HTO libraries.

We will start by showing the three most relevant metrics (number of reads, estimated number of recovered cells, fraction of reads in cells, mean reads per cell, fraction of reads mapped to exonic reads, and median genes per cell) obtained by cellranger for each of the working libraries. This information will give us an idea of the quality of the experiment as well as the sequencing and the mapping steps.

```{r fig.width=14}
table_cellranger_metrics_hashing(metrics_gex_df = metrics_gex_df,
                                 cellranger_version = "CellRanger v6.1.1")

# Keep only hashed samples
metrics_gex_df_hashed <- subset(metrics_gex_df, subset = hashing == "hashed")
```

## Sequencing QC
First, we will assess the quality of the sequenced libraries before the mapping step, as previously explained.
```{r fig.height = 6, fig.width = 14}
qc_seq_vars <- c("Antibody_Q30_Bases_in_Barcode",
                 "Antibody_Q30_Bases_in_Antibody_Read",
                 "Antibody_Q30_Bases_in_UMI")

gg_qc_seq <- purrr::map(qc_seq_vars, function(var) {
  ggplot(metrics_gex_df_hashed, aes_string(x = "library_name", y = var, fill = "gem_id")) +
    geom_col() +
    theme_bw() +
    
    ylim(0, 100) +
    labs(title = str_replace_all(var, "_", " "),
         x = "Libraries (GEM IDs)",
         y = "Antibody Q30 Bases (%)") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 8),
          axis.text.x = element_text(hjust = 1, angle = 45),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")
})

ggarrange(plotlist = gg_qc_seq, nrow = 1, ncol = 3)
```

## Antibody application QC
```{r fig.height = 6, fig.width = 14}
ab_vars <- c("Antibody_Fraction_Antibody_Reads",
             "Antibody_Fraction_Antibody_Reads_Usable")

gg_ab <- purrr::map(ab_vars, function(var) {
  ggplot(metrics_gex_df_hashed, aes_string(x = "library_name", y = var, fill = "gem_id")) +
    geom_col() +
    theme_bw() +
    
    ylim(0, 100) +
    labs(title = str_replace_all(var, "_", " "),
         x = "Libraries (GEM IDs)",
         y = "Antibody Fraction (%)") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          axis.text.x = element_text(hjust = 1, angle = 45),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")
})

ggarrange(plotlist = gg_ab, nrow = 1, ncol = 2)
```

## Sequencing saturation and depth
It is important to test which is the sequencing saturation and depth for each library, as done before.
```{r fig.height = 6, fig.width = 14}
gg_lib_size_hto <- metrics_gex_df_hashed %>%
  mutate(Antibody_Number_of_Reads_mil = Antibody_Number_of_Reads / 1000000) %>%
  ggplot(aes(x = library_name, y = Antibody_Number_of_Reads_mil, fill = gem_id)) +
  geom_col() +
  theme_bw() +
  
  labs(title = "Antibody Library size",
    x = "Libraries (GEM IDs)", 
    y =  "Number Reads (in millions)") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
          axis.text.x = element_text(hjust = 1, angle = 45),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")

gg_qc_seq_sat_hto <- metrics_gex_df_hashed %>%
  mutate(Antibody_Sequencing_Saturation_perc = Antibody_Sequencing_Saturation / 100,
         Antibody_Mean_Reads_per_Cell_tho = Antibody_Mean_Reads_per_Cell / 1000) %>%
  ggplot(aes(x = Antibody_Mean_Reads_per_Cell_tho,
             y = Antibody_Sequencing_Saturation_perc, color = gem_id)) +
  geom_point() +
  theme_bw() +
  
  ylim(0, 1) +
  geom_text_repel(aes(label = library_name), size = 4) +
  labs(title = "Antibody Sequencing Saturation",
       x = "Mean Reads per Cell (in thousands)", 
       y = "") +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.position = "none")

cowplot::plot_grid(gg_lib_size_hto, gg_qc_seq_sat_hto,
                   ncol = 2)
```

# Session Info
```{r}
sessionInfo()
```
